cmd/ls/ls_test.go

```diff
@@ -4,44 +4,79 @@ import (
 	"bytes"
 	"io/ioutil"
 	"os"
+	"os/user"
 	"path/filepath"
 	"testing"
+	"text/template"
 )
 
 func setup(t *testing.T) (string, func()) {
 	dir, err := ioutil.TempDir("", "test")
 	if err != nil {
-		fatal(err)
+		t.Error(err)
 	}
 
-	createTempFiles(dir)
+	createTempFiles(t, dir)
 
 	return dir, func() {
 		os.RemoveAll(dir)
 	}
 }
 
-func createTempFiles(dir string) {
+func checkError(t *testing.T, err error) {
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func createTempFiles(t *testing.T, dir string) {
 	files := []string{"file with space", "f1.txt", "f2.pdf", "f3"}
 	content := []byte("temporary file's content")
 
 	for _, f := range files {
 		fileName := filepath.Join(dir, f)
 		if err := ioutil.WriteFile(fileName, content, 0444); err != nil {
-			fatal(err)
+			t.Error(err)
 		}
 	}
 }
 
+func replaceUserGroup(t *testing.T, raw string) string {
+	var buf bytes.Buffer
+	currentUser, err := user.Current()
+	checkError(t, err)
+
+	group, err := user.LookupGroupId(currentUser.Gid)
+	checkError(t, err)
+
+	data := struct {
+		User  string
+		Group string
+	}{
+		User:  currentUser.Username,
+		Group: group.Name,
+	}
+
+	tpl, err := template.New("list").Parse(raw)
+	if err := tpl.Execute(&buf, data); err != nil {
+		fatal(err)
+	}
+
+	return buf.String()
+}
+
 func TestListCurrentDir(t *testing.T) {
 	tempDir, teardown := setup(t)
 	defer teardown()
```

Comment:
Yeah, this is the kind of defer that will NOT be executed if you call os.Exit on production code:

https://i.imgflip.com/1l5yzl.jpg

;*